Source/Core/Core/PowerPC/Interpreter/Interpreter_FloatingPoint.cpp

```diff
@@ -272,12 +272,34 @@ void Interpreter::fselx(UGeckoInstruction inst)
 // PS1 is said to be undefined
 void Interpreter::frspx(UGeckoInstruction inst)  // round to single
 {
-  double b = rPS0(inst.FB);
-  double rounded = ForceSingle(b);
-  SetFI(b != rounded);
-  FPSCR.FR = fabs(rounded) > fabs(b);
-  PowerPC::UpdateFPRF(rounded);
-  rPS0(inst.FD) = rPS1(inst.FD) = rounded;
+  const double b = rPS0(inst.FB);
+  const double rounded = ForceSingle(b);
+
+  if (std::isnan(b))
+  {
+    const bool is_snan = MathUtil::IsSNAN(b);
+
+    if (is_snan)
+      SetFPException(FPSCR_VXSNAN);
+
+    if (!is_snan || FPSCR.VE == 0)
+    {
+      rPS0(inst.FD) = rounded;
+      rPS1(inst.FD) = rounded;
+      PowerPC::UpdateFPRF(b);
+    }
+
+    SetFI(0);
```

Comment:
Also, do you know the behaviour on real hardware with NaN payloads? In other words, are your sure that `b == rounded` which implies `FI` and `FR` being set to 0. Can't this payload be truncated by `ForceSingle` and therefore produces a rounded value different from the original. If so, what's the hardware behaviour regarding the `FI` and `FR` bits in that case.

Reply:
The hardware's behavior is as described in D.4.1 for the SNaN and QNaN cases (which is why I put it in the description, it literally describes the entire operation of frsp). Both cases of which set FI to zero  (along with FR)

Which is for QNaNs:

```
frD ← frB[0–34] || 0b0_0000_0000_0000_0000_0000_0000_0000
FPSCR[FPRF] ← “QNaN”
FPSCR[FR FI] ← 0b00
Done
```

and for SNaNs

```
FPSCR[VXSNAN] ← 1
If FPSCR[VE] = 0 then
    Do
        frD[0–11] ← frB[0–11]
        frD[12] ← 1
        frD[13–63] ← frB[13–34] || 0b0_0000_0000_0000_0000_0000_0000_0000
        FPSCR[FPRF] ← “QNaN”
    End
FPSCR[FR FI] ← 0b00
Done
```